pkg/provisioner/ironic/ironic.go

```diff
@@ -1310,7 +1310,8 @@ func (p *ironicProvisioner) Prepare(data provisioner.PrepareData, unprepared boo
 		// When clean failed, we need to clean host provisioning settings.
 		// If unprepared is false, means the settings aren't cleared.
 		// So we can't set the node's state to manageable, until the settings are cleared.
-		if !unprepared {
+		// Also check for late binding resources in which the settings are cleared after cleaning
+		if !unprepared || lateBinding {
```

Comment:
I think this just has the opposite problem: if a previous attempt to change only the HFS (i.e. the BMH settings are not dirty) failed and we recorded that error, then when the controller retries we'll just report this error straight away and never move back to manageable. (The controller will clear the firmware settings from the BMH, meaning unprepared will be now true, but lateBinding will also remain true so we'll always take this path.)

I'm thinking if we name this new flag `force` (for consistency with other Provisioner APIs), check `if !force` here, and pass `bmhDirty` for force:

```
prov.Prepare(prepareData, bmhDirty || hfsDirty, bmhDirty)
```

(this patch is the equivalent of passing `bmhDirty && !hfsDirty` for `force`)

Reply:
I guess my suggestion relies on us being able to distinguish between the existing firmware/RAID settings having been cleared due to an error, and them having been set to nothing due to nothing being specified in the Spec. I don't think this is the case, because otherwise it would trigger manual cleaning on every host.

Reply:
Yes the fix works to use bmhDirty for force. I was able to simulate a cleaning failure when changing the HFS settings and get the error back. I was also able to recover after fixing the situation that caused the cleaning failure:
```
bmh/ostest-worker-1 -n openshift-machine-api
NAME              STATE       CONSUMER                      ONLINE   ERROR   AGE
ostest-worker-1   preparing   ostest-qdqsj-worker-0-cww5k   false            78m
 'apply_configuration', 'abortable': False, 'priority': 0}: Redfish exception occurred. Error: Redfish BIOS apply configuration failed for node 5c2bb81b-6736-43f2-a412-1c021aa76d5a. Error: HTTP PATCH http://192.168.111.1:8000/redfish/v1/Systems/193c3aa9-1836-4575-b52c-b6359c1cd38e/BIOS/Settings returned code 500. Base.1.0.GeneralError: cannot serialize 11 (type int) Extended information: [{'@odata.type': '/redfish/v1/$metadata#Message.1.0.0.Message', 'MessageId': 'Base.1.0.GeneralError'}]
```
Thanks